Dispatch begin/abort actions when fetching a version and avoid infinite fetch/error loop

willdurand opened the pull request:
Fixes #602
Fixes #556 

---

This patch is similar to the one merged to solve https://github.com/mozilla/addons-code-manager/issues/565. I added a bunch of early returns in the `reducer` (with `log.error` calls and test cases).

bobsilverberg commented on src/reducers/versions.tsx:
```diff
@@ -735,6 +764,12 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
       const version = state.versionInfo[versionId];
+
+      if (!version) {
+        _log.error(`Version missing for versionId: ${versionId}`);
```
I wonder if this should throw an actual error, rather than just logging. It shouldn't really be possible to have called `updateSelectedPath` without having a corresponding version loaded.

kumar303 replied:
> I wonder if this should throw an actual error

I guess either way works but it would be a bit better to throw an error so that Sentry gets a stack trace. It might render an error boundary and make the app unusable but at least that would be less confusing than the user seeing no change at all from their action. Thrown errors are easier to test, too 😅

bobsilverberg commented on src/reducers/versions.tsx:
```diff
@@ -757,12 +792,19 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+      const version = state.versionInfo[versionId];
+
+      if (!version) {
+        _log.error(`Version missing for versionId: ${versionId}`);
```
I wonder if this should throw an actual error, rather than just logging. It shouldn't really be possible to have called `toggleExpandedPath` without having a corresponding version loaded.

bobsilverberg commented on src/reducers/versions.tsx:
```diff
@@ -776,8 +818,13 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
-      const expandedPaths = entries
+
+      if (!version) {
+        _log.error(`Version missing for versionId: ${versionId}`);
```
I wonder if this should throw an actual error, rather than just logging. It shouldn't really be possible to have called `expandedTree` without having a corresponding version loaded.

bobsilverberg commented on src/reducers/versions.tsx:
```diff
@@ -796,12 +843,19 @@ const reducer: Reducer<VersionsState, ActionType<typeof actions>> = (
+      const version = state.versionInfo[versionId];
+
+      if (!version) {
+        _log.error(`Version missing for versionId: ${versionId}`);
```
I wonder if this should throw an actual error, rather than just logging. It shouldn't really be possible to have called `collapseTree` without having a corresponding version loaded.

bobsilverberg commented on src/reducers/versions.spec.tsx:
```diff
@@ -121,6 +121,20 @@ describe(__filename, () => {
       );
     });
 
+    it('logs an error when updateSelectedPath is called for an unknown version', () => {
```
The four test cases added that check that an error was logged for different actions are all essentially identical, other than the action being called, so I wonder if either a helper, or an `it.each()` would be appropriate for these to reduce the duplication in this file?

bobsilverberg commented on src/reducers/versions.spec.tsx:
```diff
@@ -1618,4 +1719,94 @@ describe(__filename, () => {
+        actions.abortFetchVersion({ versionId: abortFetchVersionId }),
+      );
+
+      expect(state.versionInfo[version.id]).toEqual(expect.any(Object));
```
Why are you using `any(Object)` here rather than `createInternalVersion(version)`? You are checking that the other version wasn't set to `null` but not necessarily that it was untouched in any way.

willdurand commented:
> This looks good, thanks @willdurand. I have some questions about the tests, and also about whether we should be throwing errors instead of logging.

@bobsilverberg for a similar case, we used `log.error` instead of throwing: https://github.com/mozilla/addons-code-manager/blob/aadb04c6983341e10d56ab754403dd18093fd803/src/reducers/versions.tsx#L838, so 🤷‍♂️

willdurand commented on src/reducers/versions.spec.tsx:
```diff
@@ -1618,4 +1719,94 @@ describe(__filename, () => {
+        actions.beginFetchVersion({ versionId: beginFetchVersionId }),
+      );
+
+      expect(state.versionInfo[version.id]).toEqual(expect.any(Object));
```
The purpose of the test case isn't to assert this value, only making sure it's not `null` or `undefined`. That's why `expect.any(Object)` is used here. The exact shape of this object is covered by another test case.

bobsilverberg commented:
> @bobsilverberg for a similar case, we used `log.error` instead of throwing: https://github.com/mozilla/addons-code-manager/blob/aadb04c6983341e10d56ab754403dd18093fd803/src/reducers/versions.tsx#L838
> 
> so.. 🤷‍♂️ 

I would argue that maybe that other function should be changed too. It seems reasonable to throw an error when we encounter a case that we don't believe should ever have happened.

willdurand commented:
> I would argue that maybe that other function should be changed too. It seems reasonable to throw an error when we encounter a case that we don't believe should ever have happened.

I don't know. `log.error` will send the error to sentry, without breaking the site for the end user and, for developers, logged errors are in red in the console.

I don't have a strong opinion on that but that would break the contract of a reducer: https://redux.js.org/glossary#reducer (pureness and side-effect free).

kumar303 commented on src/reducers/versions.tsx:
```diff
@@ -356,10 +365,7 @@ export const getVersionFiles = (versions: VersionsState, versionId: number) => {
-export const getVersionInfo = (
-  versions: VersionsState,
-  versionId: number,
-): Version | void => {
```
Well, I should ask, is TypeScript still inferring the wrong type?

willdurand replied:
If I remove `as Version` in a test file (on a `getVersionInfo` call), I get this TS error:

```
Type 'Version | null | undefined' is not assignable to type 'Version'. Type 'undefined' is not assignable to type 'Version'.
```

So I guess it has been fixed either _via_ a TS upgrade or _via_ the type change in this patch.

kumar303 replied:
ok, good

willdurand commented:
@bobsilverberg I updated this patch to throw errors all the time as both @kumar303 and you think it's better (I am fine with that, I removed all the log calls to be consistent).

willdurand commented:
> I think this infinite loop is new behavior because I've definitely triggered 404s in the past.

Yep, I agree. There is a link to an error page in the `Index` component. We would have noticed the infinite loop before.

bobsilverberg commented on src/reducers/versions.spec.tsx:
```diff
@@ -121,6 +121,15 @@ describe(__filename, () => {
+          undefined,
+          actions.updateSelectedPath({ selectedPath: 'pa/th', versionId: 123 }),
+        );
+      }).toThrow();
```
In a lot of our tests we assert on the actual error message. Maybe that's overkill (and introduces a maintenance burden if the messages change), but I wanted to raise it here as an inconsistency. This same comment applies to all the `expect(...).toThrow()` tests.

kumar303 replied:
Yeah, since so many errors *could* be thrown, I think it's always helpful to assert some part of the message.
